godot-core/src/engine/gfile.rs

```diff
@@ -0,0 +1,823 @@
+/// - Makes the [`FileAccess`] exclusive to its instance, not allowing parallel
+/// reads and writes, which could introduce hard to track bugs by unpredicted
+/// cursor movement. Exclusivity also ensures that when the `GFile` moves out of
+/// scope, inner [`FileAccess`] does also, automatically closing the file.
+/// Alternatively, you can [drop()] the `GFile` to close the file manually.
+/// - Implements useful Rust traits allowing better file processing, namely
+/// [`Read`], [`BufRead`], [`Write`], [`Seek`] and making it better integrated
+/// with various tools in Rust ecosystem (eg. serialization tools).
+///
+/// Files by default are always opened with little-endian, as most files are
+/// saved as such. To switch into big-endian, use `GFile.set_big_endian()`.
+///
+/// ## [`ModeFlags`]
+///
+/// Every constructor (`open_*` associated function) intakes a `flags`
+/// argument, opening the file for different types of operations. Regardless
+/// of provided `flags`, cursor is always positioned at the beginning of the
+/// file upon opening. To adjust its position use [`Seek`]-provided methods.
+///
+/// - `ModeFlags::READ` opens the file for read operations.
+/// - `ModeFlags::WRITE` opens the file for write operations. If the file
+/// doesn't exist at provided `path`, it is created. If it exists, it is
+/// truncated after the file is closed.
+/// - `ModeFlags::READ_WRITE` opens the file for read and write operations.
+/// The file is not truncated after closing.
+/// - `ModeFlags::WRITE_READ` opens the file for read and write operations.
+/// If the file doesn't exist at provided `path`, it is created. If it exists,
+/// it is truncated after the file is closed.
+///
+/// ## Examples
+///
+/// ```no_run
+/// use godot::engine::GFile;
+/// use godot::engine::file_access::ModeFlags;
+/// use godot::builtin::GString;
+///
+/// // Open file in write mode
+/// let mut my_file = GFile::open("user://save_game.sav", ModeFlags::WRITE)
+///     .expect("couldn't open file for writing");
+///
+/// // Write some lines into it
+/// my_file.write_gstring_line("This is my saved game")
+///     .expect("couldn't write first line");
+/// my_file.write_gstring_line("I played for 5 minutes")
+///     .expect("couldn't write second line");
+///
+/// // Consume object and close the file
+/// drop(my_file);
+///
+/// // Open file in read mode
+/// let mut my_file = GFile::open("user://save_game.sav", ModeFlags::READ)
+///     .expect("couldn't open file for reading");
+///
+/// // Read lines
+/// let first_line = my_file.read_gstring_line()
+///     .expect("couldn't read first line");
+/// assert_eq!(first_line, GString::from("This is my saved game"));
+/// let second_line = my_file.read_gstring_line()
+///     .expect("couldn't read second line");
+/// assert_eq!(second_line, GString::from("I played for 5 minutes"));
+///
+/// // Consume object and close the file
+/// drop(my_file);
+/// ```
```

Comment:
What if instead of all these `expect` statements, you use `?` inside two functions `save_game()` and `load_game()`?
That would:
* make the code much more readable
* only show the relevant parts (namely the `GFile` API)
* make explicit `drop` unnecessary, that's rather unusual in Rust anyway